test/unit/org/apache/cassandra/index/sai/disk/v1/trie/TrieTermsDictionaryTest.java

```diff
@@ -78,12 +80,163 @@ private void doTestExactMatch() throws Exception
+            assertEquals(NOT_FOUND, reader.ceiling(primaryKey("ba", ByteSource.LT_NEXT_COMPONENT)));
+
+            // Validate token and partition key only searches
+            assertEquals(0, reader.ceiling(primaryKey("a", "b", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(0, reader.ceiling(primaryKey("ab", "b", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(2, reader.ceiling(primaryKey("ab", "ce", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(4, reader.ceiling(primaryKey("ab", "cee", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(NOT_FOUND, reader.ceiling(primaryKey("ab", "d", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(NOT_FOUND, reader.ceiling(primaryKey("abb", "a", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(0, reader.ceiling(primaryKey("aa", "d", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(NOT_FOUND, reader.ceiling(primaryKey("abc", "a", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(NOT_FOUND, reader.ceiling(primaryKey("ba", "a", ByteSource.LT_NEXT_COMPONENT)));
+
+
+            // Validate token, partition key, and clustring column searches
+            assertEquals(0, reader.ceiling(primaryKey("a", "b", "c", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(1, reader.ceiling(primaryKey("ab", "cdd", "a", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(1, reader.ceiling(primaryKey("ab", "cde", "a", ByteSource.LT_NEXT_COMPONENT)));
+            assertEquals(2, reader.ceiling(primaryKey("ab", "cde", "z", ByteSource.LT_NEXT_COMPONENT)));
+        }
+    }
+
+    @Test
+    public void testFloor() throws Exception
+    {
+        long fp;
+        try (TrieTermsDictionaryWriter writer = new TrieTermsDictionaryWriter(indexDescriptor, indexContext, false))
+        {
+            writer.add(asByteComparable("ab"), 0);
+            writer.add(asByteComparable("abb"), 1);
+            writer.add(asByteComparable("abc"), 2);
+            writer.add(asByteComparable("abcd"), 3);
+            writer.add(asByteComparable("abd"), 4);
+            writer.add(asByteComparable("ca"), 5);
+            fp = writer.complete(new MutableLong());
+        }
+
+        try (FileHandle input = indexDescriptor.createPerIndexFileHandle(IndexComponent.TERMS_DATA, indexContext);
+             TrieTermsDictionaryReader reader = new TrieTermsDictionaryReader(input.instantiateRebufferer(), fp))
+        {
+            assertEquals(NOT_FOUND, reader.floor(asByteComparable("a")));
+            assertEquals(5, reader.floor(asByteComparable("z")));
+            assertEquals(0, reader.floor(asByteComparable("ab")));
+            assertEquals(2, reader.floor(asByteComparable("abc")));
+            assertEquals(1, reader.floor(asByteComparable("abca")));
+            assertEquals(1, reader.floor(asByteComparable("abb")));
+            assertEquals(NOT_FOUND, reader.floor(asByteComparable("abba")));
```

Comment:
Why does `abba` return `NOT_FOUND` when `abca` above returns a value?

Reply:
Great observation and question! I think we'll need to check with @blambov on the right definition.

Essentially, this comes down to how we want to handle prefix matches in the definition of floor and ceiling. Currently, prefixes are ignored in both. However, when comparing two arrays, if one is a prefix of the other, the prefix is considered less than the longer array. As such, it seems like floor should take into account the prefix while ceiling should not because floor produces a term less than or equal to a value while ceiling produces a term greater than or equal to a value.

I will work on adding a commit to show the likely fix.

Reply:
I just pushed up the commit. I found an interesting detail while writing the fix. We were encoding the bytes incorrectly when using `ByteComparable.fixedLength`, and that encoding mattered for this edge case on prefix matches. This makes me wonder if we're using it correctly elsewhere in SAI.

Reply:
> This makes me wonder if we're using it correctly elsewhere in SAI.

There are two phases. When we store bytes in an in memory trie, we leave the encodings (like the 0s and escaped 0s), but then when we serialize the trie to write to disk, we remove them here:

https://github.com/datastax/cassandra/blob/49a51690caca180fc33bd49c7b95cfc8ee0627f3/src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java#L153